### DISCUSSION TITLE
Close weboscket-connections from server-side (C++)

### DISCUSSION THREAD

**Discussion by fabianskysails** [uNetworking/uWebSockets#1528] [Q&A] [answered]
Hello! I'm trying to shut down the websocket-server correctly. Is it possible to close/abort all websocket-connections from server-side? The applications hangs on joining the websocket-thread and the following code does not close the application "quite fast" and completely:
```
loopToShutdown->defer([this]() {
    us_listen_socket_close(0, listenSocket);
  });
```
I've checked all discussions and docs but I didn't find a way how to close all connections from server-side for example in case a system need to reboot.

**Comment by uNetworkingAB** [comment, marked as answer]
It's a bit silly but you have to track your connections yourself. Something like this will make it shut down properly (this is an aera that needs improvement - ideally it should be app.close()):

std::unordered_set<void *> connectionSet;

// Track your connections
app.filter([&connectionSet](auto *s, int diff){
if (diff > 0) {
connectionSet.insert((void *) s);
} else {
connectionSet.erase((void *) s);
}
})

// When you want to close:
for (auto *s : connectionSet) {
us_socket_close(SSL, (struct us_socket_t *) s);
}

**Reply by fabianskysails** [reply]
Thank you very much for the very fast response. Looks like this code works fine for me 👍 
I already tracked the connections but I guess I did it wrong before...

**Comment by drok** [comment]
Hoping this will save someone some trouble:

The accepted solution is wrong, as it creates a use-after-free issue. When `us_socket_close()` is called, `connectionSet.erase()` removes the element from the set, also invalidating the iterator of the for loop.

Also, this solution ignores what happens with in-progress transfers. You likely want to allow those transfers to complete, or have some controlled way to close the sockets on timeouts. To do this, you should only `us_socket_close` connections where `hasResponded()` is true, and are idle.

I suggest the following is a cleaner way to end responses:

``` c++
    /* The exit code tells any waiting handler why the socket was closed. */
    #define APP_EXIT 100
    for (auto it = connectionSet.begin(); it != connectionSet.end();) {
	/* The unresponded sockets will be closed by the respective handlers after responding
          *  make sure to make the last write res->end("blah", true), where "true" means closeConnection
          *  when transfer is complete.
          */
	if ((*it)->hasResponded())
            /* advance the iterator to the next position before erase() the current */
	    us_socket_close(false, (us_socket_t*) *it++, APP_EXIT, nullptr);
        else
            it++;
    }
```
